clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
Removing last element from waiters may be wrong -- for example, some other conditions may be added to waters before timeout. We probably need to iterate through waiters to remove this condition.